### ISSUE TITLE
writing code.py causes: OSError: [Errno 5] Input/output error

### ISSUE THREAD

**Issue by dhalbert** [circuitpython/web-editor#229]
### CircuitPython Version

```python
8.2.0
9.0.0
9.1.1
all of above tested on two RP2040 boards

also reproduced the problem on 9.1.1 on an ESP32-S3

did not reproduce on a Trinket M0
```

### Workflow(s)

USB in code-beta.circuitpython.org on 2024-08-07, 2019 ET

### Browser(s)

```python
Version 127.0.6533.99 (Official Build) (64-bit) on Ubuntu 24.04
```

### Console Log

```python
Load different workflow
index.js:91 Requesting any serial device...
index.js:91 > Requested  SerialPort {onconnect: null, ondisconnect: null, readable: null, writable: null}
index.js:91 switch to SerialPort {onconnect: null, ondisconnect: null, readable: null, writable: null}
index.js:93 Read Loop Init
index.js:93 Read Loop Started
index.js:91 Getting Device UID...
index.js:93 Device UID: F6254209F45633
index.js:91 Successfully loaded directory: CIRCUITPY
index.js:91 New folder name: CIRCUITPY
index.js:91 SerialPort {onconnect: null, ondisconnect: null, readable: ReadableStream, writable: WritableStream}
index.js:89 Connected!
index.js:89 Path: /code.py
index.js:93 Current File Changed to: /code.py

[no output when writing code.py]
```

### Steps

1. Shift reload https://code-beta.circuitpython.org
2. select USB and choose serial and CIRCUITPY for an RP2040 board (I've tried two). The board has previously had `storage.erase_filesystem()` done, so CIRCUITPY is in good shape.
3. open Serial window
4. edit `code.py` and another another `print()` statement. Note that saving code.py does NOT cause an auto-reload
5. Type ctrl-D in serial window. I get:
```
>>> 
soft reboot

Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
OSError: [Errno 5] Input/output error

Code done running.

Press any key to enter the REPL. Use CTRL-D to reload.
```
The `OSError` is new. It did not show up if I did ctrl-D before editing the file.

### Additional information

**This does _not_ seem to reflect a corrupted CIRCUITPY.** Instead, it is some other I/O error, maybe due to something being done via serial. If I disconnect from web-editor and just connect via a terminal program, without resetting, and do a ctrl-D, I do not see the `OSError`.

**Comment by makermelissa** [comment]
I wasn't able to exactly fix the issue, but did find a similar issue in the process and have slowly been working it out of the system. After that, I'll try the exact reproduction steps.

**Comment by makermelissa** [comment]
Oh, I just realized it didn't happen on all boards. Will try on a RP2040 after I get the issue I notice worked out.

**Comment by makermelissa** [comment]
Just tested on the Feather RP2040 running CP 9.1.1 on my system locally and did not experience the error, so perhaps it's fixed now. After I finish #230, I'll submit a PR.

**Comment by makermelissa** [comment]
This should now be fixed via #231.

**Comment by FoamyGuy** [comment]
I'm able to re-create this issue reliably as well.

Browser: `Version 127.0.6533.99 (Official Build) (64-bit)`
OS: Ubunty 24.04

I've seen the same issue with a MEMENTO 9.1.1 and a Feather RP2040 9.0.4 and 9.1.1, it doesn't appear to be tied to a specific device or version that I can tell.

Here is a screen recording of one of occurrences on the MEMENTO
[Screencast from 2024-08-12 17-37-05.webm](https://github.com/user-attachments/assets/e50030fc-e656-4a0d-82b9-a40f3daf62e5)

I found that when the device is in the state where it's always getting the Input/Output error I can press the disconnect button at the top, and then re-connect and upon re-connecting open the serial and use ctrl-c / ctrl-d to run code.py and it does execute without the error as long as I do not change the code again before running. I have noticed what I think is an unrelated issue where disconnect / re-connecting to a device without reloading the page causes serial input and outputs to be duplicated, i'll create a seperate issue for that. 

Since it works after disconnect / reconnecting it seems to me that this issue is tied specifically to when it saves the code.py content and then tries to run it afterward. The actual content of the file seems to get saved correctly because it works without any further changes after the disconnect / reconnect.

**Comment by makermelissa** [comment]
I wonder if the issue is only appearing on linux. I'm trying with Mac OS, but have a linux computer (running Linux Mint) I can try it on. I believe Dan was probably also trying on Linux.

**Comment by dhalbert** [comment]
Yes, I only tested on Linux.

**Comment by makermelissa** [comment]
First of all, it worked fine for me on linux. However, I may have an idea about what is happening. For the RP2040, there should be a CIRCUITPY drive that you need to grant permission to. If it is trying to go through the REPL (because maybe it detects the file system isn't read only), then I think that's why it is encountering the error.

Getting down a bit deeper, I remembered that in order to make it Micropython compatible, I have it attempting to check if the file system is read only and if there are any errors it assumes that it isn't read-only. I think I need to make it more specific to an import error.

**Comment by makermelissa** [comment]
Ok, am trying the memento and noticing weirdness (code timeouts, slowness, not connecting with File System API, etc). Hopefully getting it working will iron out the issue. I think the weirdness is causing it to think it should do file ops through the REPL and then it can't write to a read-only file system.

**Comment by makermelissa** [comment]
Ok, good news is same board works much better after trying on my local system (which only assumes not read-only on import error instead of all errors). I think I can improve the connection speed/weirdness though and can call this good.

**Comment by makermelissa** [comment]
I just noticed in Dan's log:
```
index.js:91 New folder name: CIRCUITPY
```

This means that it is connecting through the File System API, which is correct. So, I don't think my hypothesis is correct.

**Comment by makermelissa** [comment]
I removed the beta label because I don't think this has quite been fixed on beta yet.

**Comment by makermelissa** [comment]
I think when trying to reproduce this before, I had missed the step where you typed Ctrl-D. Now after following the instructions above, I get:
```
soft reboot

Auto-reload is off.
Running in safe mode! Not running saved code.

You are in safe mode because:
CircuitPython core code crashed hard. Whoops!
Hard fault: memory access or instruction error.
Please file an issue with your program at github.com/adafruit/circuitpython/issues.
Press reset to exit safe mode.

Press any key to enter the REPL. Use CTRL-D to reload.
```

This was on the Metro RP2350. This could actually be a CircuitPython bug. My suspicion is that maybe it does this after being put in a certain mode. I'll need to take a closer look at some of the underlying REPL module code.

**Comment by makermelissa** [comment]
At some point, it does run some code to check if the filesystem is readonly. We may need to put it in normal mode if it is readonly and hopefully that will solve it from this end. However, if I can fiigure out how to reproduce the bug manually, this could also be fixed in CP.

**Comment by makermelissa** [comment]
Ok, I couldn't reproduce by hand. Then I gave it another try and it worked as expected. I'll try this on an RP2040 feather next.

**Comment by makermelissa** [comment]
I loaded CP 9.1.1 onto a RP2040 board and wasn't able to reproduce the issue. When one of you gets a chance, could you verify this is still an issue? I've made several bug fixes to circuitpython-repl-js and I'm hoping those fixed the bug. Please try with latest firmware as well in case this got fixed there.

**Comment by makermelissa** [comment]
Since this issue came up more with linux computers, it makes sense to test this on a linux computer. If I am unable to reproduce the issue, we may need to close it until we find a reliable way to recreate the conditions of the error. I will give it a try using 9.1.1 to remove that as a variable. If I can reproduce the issue, then I will try on a more recent version of CP.

**Comment by makermelissa** [comment]
I was finally able to reproduce the issue on linux and CP 9.1.1 on a Feather RP2040.

**Comment by makermelissa** [comment]
The issue is still present in 9.2.8.

**Comment by makermelissa-ai-assistant** [comment]
## Root cause analysis

Spent some time digging into this and wanted to leave notes for posterity, since the symptom ("OSError after Ctrl-D from web editor") is misleading about where the bug actually lives.

### Where the bug lives

**Not in the web editor. Not in CircuitPython firmware. In the Linux host's filesystem layer.**

- macOS and Windows: do **not** reproduce.
- Linux: reproduces consistently on the USB MSC workflow.
- Network workflow (web/BLE): unaffected on all OSes.

### The mechanism

1. The web editor uses the Chromium **File System Access API** to write `code.py` (and other files) to the CIRCUITPY drive. The browser does a normal `write()` + `close()` against the host filesystem.
2. On Linux, CIRCUITPY is auto-mounted as `vfat` with the **default async** behavior. Writes land in the kernel **page cache** and are flushed to the device on the kernel's writeback timer — typically up to ~30 seconds later.
3. Pressing <kbd>Ctrl</kbd>+<kbd>D</kbd> over serial triggers an immediate soft-reboot on the device. CircuitPython tries to import `code.py` from the device's view of the FAT filesystem.
4. If the host hasn't flushed yet, CircuitPython sees an **inconsistent on-device filesystem** — the directory entry, FAT chain, and/or data sectors don't agree — and the read fails with `OSError: [Errno 5] Input/output error`.
5. ~30 s later, the kernel finally flushes; the next reboot works fine. The data was never lost — just delayed.

### Why this only became visible recently

The web editor's USB workflow (FSAPI) writes through the host filesystem. The older REPL workflow writes via serial directly to the device, bypassing the host filesystem entirely — so it doesn't hit this race.

### Why I didn't fix it in JS

I tried, in this order:
1. `createWritable({ mode: "exclusive" })` — does not change writeback timing.
2. Post-write file re-open + small delay — no effect.
3. Synthetic "nudge" reads/writes (re-read, sentinel write, re-touch, 1 MB bulk write) to try to push the writeback timer — only sometimes works, and even when it does, the per-save UX cost is 11–22 s.
4. A "warm-up flush" on first save per session — fixed first-save failures in a diagnostic build but production was still failing every save (cold pipeline).

The fundamental problem: **browsers do not expose `fsync()` or any direct way to force the kernel page cache to flush.** The File System Access API is layered on top of normal filesystem writes, and the kernel decides when to flush. Any workaround is a heuristic against kernel writeback timing, not a fix.

### Verification of the host-OS root cause

Remounting CIRCUITPY with the `sync` option:

```sh
udisksctl unmount -b /dev/sdX1
udisksctl mount -b /dev/sdX1 -o sync
```

→ **3/3 trials, no OSError.** Restoring the default mount → bug returns. This is conclusive: the bug is the kernel writeback delay, and synchronous mount is the deterministic fix.

### The fix in #491

Rather than ship a fragile JS workaround, #491:

- Adds a Linux-only collapsed notice in the USB connect dialog explaining the issue with the `udisksctl` commands.
- Adds [`docs/LINUX_USB_MOUNT.md`](https://github.com/circuitpython/web-editor/blob/main/docs/LINUX_USB_MOUNT.md) with a permanent `udev` rule that matches by `ID_FS_LABEL=="CIRCUITPY"` and applies `UDISKS_MOUNT_OPTIONS+="sync"` so every CircuitPython board automounts correctly going forward.
- Does **not** touch `fsapi-file-transfer.js` or the read-only fallback path (#327 preserved).

### Trade-off note

`-o sync` makes every write to CIRCUITPY block until the device acknowledges it. For interactive editing (small saves) this is invisible. For copying large files (firmware, libraries) it will be slower than async — but the data is reliably on the device when the copy returns, which is generally what you want anyway for a flash-based device.

### Useful diagnostic commands (for anyone hitting this later)

```sh
# See current mount options:
mount | grep CIRCUITPY

# See live block-device write activity (look for delayed bursts vs prompt writes):
watch -n 0.5 cat /sys/block/sdX/sdX1/stat

# One-shot fix for current session:
udisksctl unmount -b /dev/sdX1
udisksctl mount -b /dev/sdX1 -o sync
```

**Comment by makermelissa-ai-assistant** [comment]
Reopening — the original fix (#491) was reverted in #493. Exploring an alternative approach: on Linux, skip the FSAPI path entirely and use the REPL-based file transfer client instead (since the host kernel page cache is the root cause and the File System Access API doesn't expose fsync). User will be prompted to eject CIRCUITPY so the firmware filesystem becomes writable. New PR coming.

**Comment by makermelissa-ai-assistant** [comment]
Following up on the revert in #493 and @dhalbert's suggestion above — exploring an alternative approach:

On Linux, skip the FSAPI path entirely and use the existing REPL-based file transfer client instead (since the host kernel page cache is the root cause, and the File System Access API doesn't expose fsync). The user will be prompted to eject CIRCUITPY so the firmware filesystem becomes writable, then everything else flows through the same REPL path that boards-without-CIRCUITPY already use.

PR incoming. Could a maintainer reopen this issue so the new PR can close it cleanly on merge?
